extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,730 @@
+          break;
+        case PREVIEW_ACTIONS.TOGGLE_DEVICE_CHIP:
+          this.onAddDeviceChipToggled_(actionElement);
+          break;
+      }
+    }
+  }
+
+  /**
+   * @private
+   * @param {string} deviceName
+   */
+  addDevice_(deviceName) {
+    const deviceSpecs = ALL_DEVICES.find((d) => d.name === deviceName);
+    if (!deviceSpecs) {
+      return;
+    }
+    const deviceLayout = this.buildDeviceLayout_(deviceSpecs, this.storyUrl_);
+    deviceSpecs.element = deviceLayout;
+    deviceSpecs.chip = this.buildDeviceChip_(deviceSpecs.name);
+    this.mutateElement(() => {
+      this.element.appendChild(deviceLayout);
+      this.element
+        .querySelector('.i-amphtml-story-dev-tools-device-chips')
+        .appendChild(deviceSpecs.chip);
+    });
+    this.devices_.push(deviceSpecs);
+    this.updateDevicesInHash_();
+  }
+
+  /**
+   * Try to remove a device with the name passed.
+   * @private
+   * @param {string} deviceName
+   * @return {bool} whether a device was found and removed with that name.
+   */
+  removeDevice_(deviceName) {
+    const device = this.devices_.find((d) => d.name === deviceName);
+    if (device) {
+      this.mutateElement(() => {
+        device.chip.remove();
+        device.element.remove();
+      });
+      this.devices_ = this.devices_.filter((d) => d != device);
+      this.updateDevicesInHash_();
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @param {string} deviceName
+   * @return {!Element} the device chip
+   */
+  buildDeviceChip_(deviceName) {
+    const deviceChip = buildDeviceChipTemplate(this.element);
+    deviceChip.querySelector(
+      '.i-amphtml-story-dev-tools-device-chip-name'
+    ).textContent = deviceName;
+    deviceChip.setAttribute('data-action', PREVIEW_ACTIONS.REMOVE_DEVICE);
+    deviceChip.setAttribute('data-device', deviceName);
+    return deviceChip;
+  }
+
+  /**
+   * @param {!Element} chipElement
+   * @private
+   */
+  onAddDeviceChipToggled_(chipElement) {
+    const deviceName = chipElement.getAttribute('data-device');
+    if (this.removeDevice_(deviceName)) {
+      chipElement.setAttribute('inactive', '');
+    } else {
+      chipElement.removeAttribute('inactive');
+      this.addDevice_(deviceName);
+    }
+  }
+
+  /**
+   * If the devices are not the default ones, add them to the hashString
+   * @private
+   */
+  updateDevicesInHash_() {
+    const devicesStringRepresentation = this.devices_
+      .map((device) => simplifyDeviceName(device.name))
+      .join(';');
+    this.element.setAttribute('devices', devicesStringRepresentation);
+    updateHash(
+      {
+        'devices':
+          devicesStringRepresentation != DEFAULT_DEVICES
+            ? devicesStringRepresentation
+            : null,
+      },
+      this.win
+    );
+  }
+
+  /**
+   * When measure changes, we recalculate the positions of the devices to keep them spaced and scaled evenly.
+   * @override
+   * */
+  onLayoutMeasure() {
```

Comment:
Are you sure you want `onLayoutMeasure` that is triggered A LOT, versus `onMeasuredChange` that is only triggered when needed?

If you're keeping `onLayoutMeasure`, at least write an early return for when the layoutBox didn't change.

Reply:
I think you pointed out that layoutMeasure triggers when you do a mutateElement (eg when you add a device) which is useful because you need to recalculate the layout in that case. I can cache the measures and check before mutating though!

Reply:
If that's what you're trying to accomplish you should manually call a private method when you add a device, and on resize (onMeasuredChange). `onLayoutMeasure` is called way too often